### PR TITLE
[FSDP][2/N] Refactor state dict hook registration

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#90777 [FSDP][2/N] Refactor state dict hook registration**
* #90767 [FSDP][1/N] `fully_shard` state dict
* #90660 [FSDP] Clean up `FlatParamHandle` dtypes, post-backward hook

This PR includes some follow-ups from the previous PR to clean up the code.

### CODE REVIEW

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 1]
```diff
@@ -706,55 +706,40 @@ def _post_load_state_dict_hook(
...
-def _register_load_state_dict_post_hooks(state: _FSDPState) -> None:
-    """Registers the post-load state dict hooks."""
+def _register_state_dict_hooks_base(
+    state: _FSDPState,
+    hook_registration_fn_name: str,
+    hook: Callable,
+    hook_registration_fn_kwargs: Dict[str, Any],
+) -> None:
+    """Registers ``hook`` using ``hook_registration_fn``."""
+    hook_with_state = functools.partial(hook, state)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_state_dict_utils.py:736]: nit: maybe add a TODO and follow up issue to get state directly from module and avoid this partial here. cc @fegin

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks for addressing the feedback!

File: torch/distributed/fsdp/_state_dict_utils.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/_state_dict_utils.py:736]: Done.

**Pull request closed without merging**